scenario/constraints/ProposalConstraint.ts

```diff
@@ -1,45 +1,144 @@
+    // Fund voters account with eth for sending transactions
+    // COMP biggest whales
+    const voters = [
+      '0xea6c3db2e7fca00ea9d7211a03e83f568fc13bf7',
+      '0x61258f12c459984f32b83c86a6cc10aa339396de',
+      '0x9aa835bc7b8ce13b9b0c9764a52fbf71ac62ccf1',
+      '0x683a4f9915d6216f73d6df50151725036bd26c02',
+      '0xa1b61405791170833070c0ea61ed28728a840241',
+      '0x88fb3d509fc49b515bfeb04e23f53ba339563981',
+      '0x8169522c2c57883e8ef80c498aab7820da539806'
+    ];
+    await fundVoters(world, voters, context.primaryActor());
+    const pendingProposals = await getAllPendingProposals(world, governor.address, voters[0]);
+
+    for (let proposal of pendingProposals) {
       solutions.push(async function (context: T): Promise<T> {
-        // XXX queue/execute each proposal in the list
-        // XXX if gov chain is not local chain, simulate bridge
-        debug(`XXX...`);
-        return context;
+        try {
+          // XXX if gov chain is not local chain, simulate bridge
+          debug(`Processing pending proposal ${proposal.proposalId}`);
+          const { proposalId, startBlock, endBlock } = proposal;
+          const blockNow = await world.hre.ethers.provider.getBlockNumber();
+          const blocksUntilStart = startBlock - blockNow;
+          const blocksFromStartToEnd = endBlock - Math.max(startBlock, blockNow);
+          if (blocksUntilStart > 0) {
+            await context.mineBlocks(blocksUntilStart);
+          }
+
+          for (const voter of voters) {
+            try {
+              // Voting can fail if voter has already voted
+              const voterSigner = await world.impersonateAddress(voter);
+              const govAsVoter = await world.hre.ethers.getContractAt(
+                'IGovernorBravo',
+                governor.address,
+                voterSigner
+              ) as IGovernorBravo;
+              await govAsVoter.castVote(proposalId, 1);
+            } catch (err) {
+              debug(`Error while voting ${err}`);
+            }
+          }
+          await context.mineBlocks(blocksFromStartToEnd);
+
+          const adminSigner = await world.impersonateAddress(voters[0]);
+          const governorAsAdmin = await world.hre.ethers.getContractAt(
+            'IGovernorBravo',
+            governor.address,
+            adminSigner
+          ) as IGovernorBravo;
+
+          // Queue proposal
+          const queueTxn = await (await governorAsAdmin.queue(proposalId)).wait();
+          const queueEvent = queueTxn.events?.find(event => event.event === 'ProposalQueued');
+          await context.setNextBlockTimestamp(queueEvent?.args?.eta.toNumber());
+
+          // Execute proposal
+          await governorAsAdmin.execute(proposalId);
+
+          debug(`Pending proposal ${proposalId} was executed`);
+          return context;
+        } catch (err) {
+          debug(`Failed with error ${err}`);
+          return context;
+        }
```

Comment:
Look good!

Reply:
Wondering if there might be a way to break this down into a bit more composable units 🤔

Reply:
Since most of the logic in this for-loop seems to be the same as `context.fastGovernanceExecute(calldata)`, we could create a version of `context.fastGovernanceExecute(proposalId)` that takes a proposalId as an argument.

The two versions of `context.fastGovernanceExecute` will pretty much share the same code and we won't have to duplicate the logic here.

Reply:
agree, I actually working on such a version right now. So that `fastGovernanceExecute` and proposal constraint reuse the same code.